Prepare 2.1.0 release

eliarbel opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary
This PR prepares the 2.1.0 release.

### Details and comments

- [x] Set the version number to 2.1.0
- [x] Fix typos, broken links and copyedit the 2.1 release notes 
- [x] Move any loose release notes into the releasenotes/notes/2.1 directory
- [x] Add a prelude release note for the release

eliarbel committed:
Version update and first pass at the release notes

beckykd commented on releasenotes/notes/2.1/Added-random-circuit-from-graph-95c22eeabdea89d0.yaml:
```diff
@@ -1,20 +1,20 @@
   - |
-    Added a function :func:`~qiskit.circuit.random.utils.random_circuit_from_graph` that generates a random circuit that 
+    Added a function :func:`~qiskit.circuit.random.random_circuit_from_graph` that generates a random circuit that
     induces the same interaction graph as the interaction graph specified by `interaction_graph`.
```
```suggestion
    induces the same interaction graph as the one specified by `interaction_graph`.
```

beckykd commented on releasenotes/notes/2.1/Added-random-circuit-from-graph-95c22eeabdea89d0.yaml:
```diff
@@ -1,20 +1,20 @@
-    each edge is drawn uniformly, i.e. each two-qubit gate represented by an edge in the interaction graph has the 
-    same probability of getting added to the random circuit. If only a subset of edge probabilities are set, 
+    The probability of randomly drawing an edge from the interaction graph as a two-qubit gate can be set by the
+    user in the weight attribute of an edge in the input interaction graph. If the user does not set the probability,
```
```suggestion
    user in an edge's weight attribute in the input interaction graph. If the user does not set the probability,
```

beckykd commented on releasenotes/notes/2.1/Added-random-circuit-from-graph-95c22eeabdea89d0.yaml:
```diff
@@ -1,20 +1,20 @@
-    same probability of getting added to the random circuit. If only a subset of edge probabilities are set, 
+    The probability of randomly drawing an edge from the interaction graph as a two-qubit gate can be set by the
+    user in the weight attribute of an edge in the input interaction graph. If the user does not set the probability,
+    each edge is drawn uniformly, i.e. each two-qubit gate represented by an edge in the interaction graph has the
```
```suggestion
    each edge is drawn uniformly.  That is, each two-qubit gate represented by an edge in the interaction graph has the
```

beckykd commented on releasenotes/notes/2.1/Added-random-circuit-from-graph-95c22eeabdea89d0.yaml:
```diff
@@ -1,20 +1,20 @@
+    same probability of getting added to the random circuit. If only a subset of edge probabilities are set,
     `ValueError` will be raised.
 
     This is an example where 'cp_map' is a list of edges with some arbitrary weights.
```
```suggestion
    In this example, 'cp_map' is a list of edges with arbitrary weights.
```

beckykd commented on releasenotes/notes/2.1/add-clifford-t-transpiler-pipeline-e640b5ba64eab13a.yaml:
```diff
@@ -9,7 +9,7 @@ features_transpiler:
     :func:`.get_clifford_gate_names`.
```
I can't comment on the above lines directly, but I'd tweak the wording a bit: 

...basis set consists only of Clifford+T gates.  Formally, a Clifford+T basis set must contain only Clifford gates, along with a :class:`.TGate`,  :class:`.TdgGate`, or both. The full list of supported Clifford gates can be obtained by using...

beckykd commented on releasenotes/notes/2.1/add-clifford-t-transpiler-pipeline-e640b5ba64eab13a.yaml:
```diff
@@ -32,8 +32,8 @@ features_transpiler:
     It is important to note that the specified Clifford+T basis gate set should be
-    universal, or else transpilation might not succeed. While the gate set 
-    ``["h", "t", "tdg"]``or even ``["h", "t"]`` is sufficient for universality,
+    universal, or else transpilation might not succeed. While the gate set
```
Something else I can't comment on directly, but would suggest: 

"Previously, the generated pass manager could not handle the example above because it couldn't decompose single-qubit :class:`.UGate` rotation gates into Clifford+T gates.  However, the new pass manager uses the Solovay-Kitaev decomposition to approximate single-qubit rotation gates
     by using ``H``, ``T``, and ``Tdg`` gates, and calls the :class:`~.BasisTranslator`"...

beckykd commented on releasenotes/notes/2.1/add-clifford-t-transpiler-pipeline-e640b5ba64eab13a.yaml:
```diff
@@ -32,8 +32,8 @@ features_transpiler:
+    universal, or else transpilation might not succeed. While the gate set
+    ``["h", "t", "tdg"]`` or even ``["h", "t"]`` is sufficient for universality,
     it is recommended to add more Clifford gates to the set if possible, as
     otherwise the translation might be less efficient. For example, not including
```
```suggestion
    otherwise the translation might be less efficient. For example, if S-gate is not included, 
```

beckykd commented on releasenotes/notes/2.1/add-clifford-t-transpiler-pipeline-e640b5ba64eab13a.yaml:
```diff
@@ -32,8 +32,8 @@ features_transpiler:
+    ``["h", "t", "tdg"]`` or even ``["h", "t"]`` is sufficient for universality,
     it is recommended to add more Clifford gates to the set if possible, as
     otherwise the translation might be less efficient. For example, not including
     S-gate might trigger decomposing S-gates into pairs of T-gates (that is,
```
```suggestion
    S-gates might be decomposed into pairs of T-gates (that is,
```

beckykd replied:
(that is, Clifford gates might be decomposed into non-Clifford...

beckykd replied:
Suggest: 

"Following is a slightly larger example::"

beckykd commented on releasenotes/notes/2.1/add-new-half-adder-6532be58252b2a86.yaml:
```diff
@@ -10,15 +10,15 @@ features_transpiler:
     Added new high-level-synthesis plugin for synthesizing a :class:`.HalfAdderGate`:
 
-    * :class:`.HalfAdderSynthesisR25`, based on :func:`.adder_ripple_rv_25`.
+    * :class:`.HalfAdderSynthesisR25`, based on :func:`.adder_ripple_r25`.
```
Suggestion: 

"Added a new .." (there are a couple of places where the "a" has been left out.)

beckykd commented on releasenotes/notes/2.1/add-new-half-adder-6532be58252b2a86.yaml:
```diff
@@ -10,15 +10,15 @@ features_transpiler:
     The :class:`.HalfAdderSynthesisDefault` has also been updated to follow the following sequence
-    of half adder synthesizers: ``"HalfAdder.ripple_r25"`` when the number of qubits is 
-    :math:`\leq 3`, ``"HalfAdder.ripple_c04"`` when 1 ancillary qubit is available, 
+    of half adder synthesizers: ``"HalfAdder.ripple_r25"`` when the number of qubits is
```
```suggestion
    of half adder synthesizers: ``"HalfAdder.ripple_r25"`` when there are 
```

beckykd commented on releasenotes/notes/2.1/add-new-half-adder-6532be58252b2a86.yaml:
```diff
@@ -10,15 +10,15 @@ features_transpiler:
-    of half adder synthesizers: ``"HalfAdder.ripple_r25"`` when the number of qubits is 
-    :math:`\leq 3`, ``"HalfAdder.ripple_c04"`` when 1 ancillary qubit is available, 
+    of half adder synthesizers: ``"HalfAdder.ripple_r25"`` when the number of qubits is
+    :math:`\leq 3`, ``"HalfAdder.ripple_c04"`` when 1 ancillary qubit is available,
```
```suggestion
    :math:`\leq 3` qubits, ``"HalfAdder.ripple_c04"`` when one ancillary qubit is available, and 
```

beckykd commented on releasenotes/notes/2.1/add-new-half-adder-6532be58252b2a86.yaml:
```diff
@@ -10,15 +10,15 @@ features_transpiler:
-    :func:`.adder_ripple_r25`. To obtain the old behaviour, use the definition of
-    :func:`.adder_qft_d00` instead.
\ No newline at end of file
+    :func:`.adder_ripple_r25`. To obtain the old behavior, use the definition of
```
```suggestion
    :func:`.adder_ripple_r25`. To obtain the old behavior, use 
```

beckykd commented on releasenotes/notes/2.1/annotation-frameworkd-d9f86b90720359b5.yaml:
```diff
@@ -1,15 +1,15 @@
+    A new module :mod:`qiskit.circuit.annotation` and principle object :class:`.Annotation`
+    have been added.
 
     Annotations are a way of tagging instructions (currently only :class:`.BoxOp`, in the initial
```
```suggestion
    Annotations are a way of tagging instructions (currently only :class:`.BoxOp`)
```

beckykd commented on releasenotes/notes/2.1/annotation-frameworkd-d9f86b90720359b5.yaml:
```diff
@@ -1,15 +1,15 @@
+    have been added.
 
     Annotations are a way of tagging instructions (currently only :class:`.BoxOp`, in the initial
     implementation) with local, user-custom data.  This data is intended to be consumed by custom
```
```suggestion
     with local, user-custom data.  This data is intended to be consumed by custom
```

beckykd commented on releasenotes/notes/2.1/c-api-version-support-55b018bf9552db1f.yaml:
```diff
@@ -2,7 +2,7 @@
 features_c:
   - |
     Added support for querying the Qiskit version information from the C API via the following macros:
```
```suggestion
    Added support for querying the Qiskit version information from the C API by using the following macros:
```

beckykd commented on releasenotes/notes/2.1/deprecate-library-circuits-4f8b234500f90d50.yaml:
```diff
@@ -2,7 +2,7 @@ deprecations_circuits:
   - |
     The circuit library underwent a refactoring in the Qiskit 1.3 release,
```
```suggestion
    The circuit library underwent a refactoring in the Qiskit v1.3 release,
```

beckykd commented on releasenotes/notes/2.1/deprecate-library-circuits-4f8b234500f90d50.yaml:
```diff
@@ -2,7 +2,7 @@ deprecations_circuits:
     The circuit library underwent a refactoring in the Qiskit 1.3 release,
     in which alternatives for objects of type :class:`.QuantumCircuit` were provided
-    that are either a :class:`.Instruction` or a Python function for construction. This refactoring
+    that are either an :class:`.Instruction` or a Python function for construction. This refactoring
```
Other comments: 

"...alternatives introduced in Qiskit v1.3...period and will not be removed until Qiskit v3.0.

beckykd replied:
"for more details, but some common circuits and their replacements follow: "

beckykd commented on releasenotes/notes/2.1/deprecate-library-circuits-4f8b234500f90d50.yaml:
```diff
@@ -2,7 +2,7 @@ deprecations_circuits:
     The circuit library underwent a refactoring in the Qiskit 1.3 release,
     in which alternatives for objects of type :class:`.QuantumCircuit` were provided
-    that are either a :class:`.Instruction` or a Python function for construction. This refactoring
+    that are either an :class:`.Instruction` or a Python function for construction. This refactoring
```
"and the migration guide for more details" -- can that link to the migration guide?

eliarbel replied:
Apparently, there is no migration guide per se so I refined the note accordingly.

beckykd commented on releasenotes/notes/2.1/extend-deprecation-tuple-inst-f24ab7a5632191e3.yaml:
```diff
@@ -4,6 +4,6 @@ deprecations_circuits:
     The deprecated tuple-like interface for :class:`~.circuit.CircuitInstruction` was not
     removed in this release as originally planned. It will be removed in Qiskit 3.0.0
```
```suggestion
    removed in this release as originally planned. It will be removed in Qiskit v3.0
```

beckykd replied:
Here, 3.0.0 was used and above it's just 3.0.  Should that be consistent?

eliarbel replied:
Having consistency here would be nice. I've updated other places where it made sense but left a few mentions explicit (e.g. 2.3.0 and not v2.3) where the extra accuracy seems to be more appropriate.

beckykd commented on releasenotes/notes/2.1/fix-assign_parameters-string-input-with-strict-false-5597010d6d0c4a5f.yaml:
```diff
@@ -0,0 +1,6 @@
+---
+fixes:
+  - |
+    Fixed a bug in method :meth:`.QuantumCircuit.assign_parameters` that results in error with string inputs
```
```suggestion
    Fixed a bug in method :meth:`.QuantumCircuit.assign_parameters` that results in an error when there are string inputs
```

beckykd replied:
I couldn't tell if string inputs should result in an error and that was the fix (which is what the above sounds like to me), or if string inputs used to result in an error and that was the fix.

eliarbel replied:
It's the latter, but I agree, it wasn't clear. I updated the phrasing.

beckykd commented on releasenotes/notes/2.1/fix-qpy-duplicated-param-names-8404ff8845dabbd5.yaml:
```diff
@@ -1,8 +1,8 @@
   - |
-    Fixed a bug in meth :class:`.qpy.load` where it could fail to deserialize
+    Fixed a bug in the :func:`.qpy.load` function where it could fail to deserialize
     circuits whose parameters had been re-assigned to parameters with the same names.
```
```suggestion
    circuits whose parameters had been reassigned to parameters with the same names.
```

beckykd commented on releasenotes/notes/2.1/improve-controlled-UnitaryGate-5f3aa3d44034cfa7.yaml:
```diff
@@ -5,5 +5,5 @@ features_synthesis:
     the :class:`.CXGate` count in the case that the input unitary happens to be a controlled unitary.
```
```suggestion
    the :class:`.CXGate` count in the case that the input unitary is a controlled unitary.
```

beckykd commented on releasenotes/notes/2.1/observables-array-7d7544b3bb3fbaa3.yaml:
```diff
@@ -2,6 +2,7 @@
 features:
   - |
     Incorporated the :class:`.SparseObservable` class with the :class:`.ObservablesArray` class.
+
```
"Estimator PUBs can now be defined by using... However, ... Paulis..."

beckykd commented on releasenotes/notes/2.1/qiskit-unitary-synthesis-plugin-8ae54f82a65a09a3.yaml:
```diff
@@ -2,17 +2,17 @@
     Added a new unitary synthesis plugin :class:`.CliffordUnitarySynthesis` that
-    attempts to syntesize a given unitary gate by checking if it can be represented
+    attempts to synthesize a given unitary gate by checking if it can be represented
     by a Clifford, in which case it returns a circuit implementing this unitary and
```
```suggestion
    by a Clifford, in which case it returns a circuit that implements this unitary and
```

beckykd commented on releasenotes/notes/2.1/qiskit-unitary-synthesis-plugin-8ae54f82a65a09a3.yaml:
```diff
@@ -2,17 +2,17 @@
-    attempts to syntesize a given unitary gate by checking if it can be represented
+    attempts to synthesize a given unitary gate by checking if it can be represented
     by a Clifford, in which case it returns a circuit implementing this unitary and
     consisting only of Clifford gates.
```
```suggestion
    consists only of Clifford gates.
```

beckykd commented on releasenotes/notes/2.1/update-cs-csdg-in-equivalence-library-c1e70d3246f4aa6d.yaml:
```diff
@@ -2,6 +2,6 @@
     The standard equivalence library has additional equivalences for :class:`.CSGate` and :class:`.CSdgGate`.
-    When transpiling to the `["cx", "u"]` basis set, the :class:`.BasisTranslator` transpiler pass now uses
+    When transpiling to the ``["cx", "u"]`` basis set, the :class:`.BasisTranslator` transpiler pass now uses
     2  :class:`.CXGate` gates instead of 3 for :class:`.CSdgGate`, and in addition reduces the number of single-qubit
```
```suggestion
    two  :class:`.CXGate` gates instead of three for :class:`.CSdgGate`, and in addition reduces the number of single-qubit
```

beckykd reviewed (commented):
Lots of good information!

eliarbel committed:
Update releasenotes/notes/2.1/Added-random-circuit-from-graph-95c22eeabdea89d0.yaml

Co-authored-by: Rebecca Dimock <66339736+beckykd@users.noreply.github.com>

eliarbel committed:
Apply comments from Becky's review

eliarbel committed:
Merge branch 'prepare-2.1.0' of github.com:eliarbel/qiskit into prepare-2.1.0

eliarbel committed:
Fix broken links and other formatting issues

eliarbel committed:
Apply fixes after proofreading

mtreinish committed:
Update feature notes

This commit goes through the new feature release notes and updates and
expands on them for clarity.

mtreinish committed:
Update other notes

mtreinish committed:
Update deprecation notes

mtreinish committed:
Update upgrade notes

mtreinish committed:
Update C API target note

eliarbel committed:
Add prelude

eliarbel committed:
Refine circuit library release note

Remove the mention of a migration guide

mtreinish commented on qiskit/circuit/__init__.py:
```diff
@@ -280,6 +280,11 @@
+.. autosummary::
+    :toctree: ../stubs/
+
+    BoxOp
```
I think with the rebase on top of https://github.com/Qiskit/qiskit/commit/414e6156138308d690cf80e658b37780da8ea7b4 this and the change in `qiskit/circuit/annotation.py` may not be needed anymore.

eliarbel replied:
If we remove these we will not have the `BoxOp` and `Annotation` classes in the index anymore. That said, I agree that `BoxOp` in the circuit module documentation appears too early. Maybe we should move the entire paragraph describing it to appear after the Control flow in circuits section?

jakelishman replied:
We will - the point is that 414e615 added them properly

mtreinish replied:
This is what's causing the docs failure CI now: https://github.com/Qiskit/qiskit/actions/runs/15736021423/job/44348814014?pr=14565#step:5:1289 there are two copies of `Annotation` in the docs now and sphinx doesn't know which to link to for inline refs.

eliarbel replied:
Right. Apparently, it was a problem in my local branch that didn't include that commit.

mtreinish commented on releasenotes/notes/2.1/Added-random-circuit-from-graph-95c22eeabdea89d0.yaml:
```diff
@@ -24,13 +24,13 @@ features_circuits:
         pydi_graph.add_edges_from(cp_map)
```
You can remove the explicit add nodes step and use: https://www.rustworkx.org/apiref/rustworkx.PyDiGraph.extend_from_weighted_edge_list.html instead here. Not that it is a big deal either way.

mtreinish commented on releasenotes/notes/2.1/add-clifford-t-transpiler-pipeline-e640b5ba64eab13a.yaml:
```diff
@@ -22,25 +21,25 @@ features_transpiler:
       qct = pm.run(qc)
       print(qct.count_ops())
```
So just pointing out that a `print()` here doesn't display anything in the docs. We can use the plot directive to include a circuit diagram, but this print won't show anything.

alexanderivrii replied:
I agree that this `print()` statement is superfluous, but the reason that I have included this in the original release notes is because I was thinking of a code example as something a user can run to get meaningful results. So here it was my way of telling the user to examine operations on the transpiled circuit and to notice how they are wonderfully in the Clifford+T basis set. But I now agree that it's probably better to delete this line.

eliarbel replied:
we can add comments to the example code with the print outputs to show the reduction

eliarbel replied:
I've added the outputs as text code blocks. Let me know if you want to remove the print statements altogether.

mtreinish commented on releasenotes/notes/2.1/add-new-half-adder-6532be58252b2a86.yaml:
```diff
@@ -1,24 +1,27 @@
-    :func:`.adder_qft_d00` instead.
\ No newline at end of file
+    The :attr:`~.HalfAdderGate.definition` attribute of the :class:`.HalfAdderGate`
+    has been changed to internall use :func:`.adder_ripple_r25` to generate the definition
```
```suggestion
    has been changed to internally use :func:`.adder_ripple_r25` to generate the definition
```

mtreinish commented on releasenotes/notes/2.1/add-new-half-adder-6532be58252b2a86.yaml:
```diff
@@ -1,24 +1,27 @@
+    Added a new high-level-synthesis plugin for synthesizing a :class:`.HalfAdderGate`:
 
-    * :class:`.HalfAdderSynthesisR25`, based on :func:`.adder_ripple_rv_25`.
+    * :class:`.HalfAdderSynthesisR25`, based on :func:`.adder_ripple_r25`.
```
Does this need to be a bulleted list, there is only a single item? 

Also for new built-in plugins it's useful to show how to use them in the release note. The plugin class by itself isn't actually user facing a user typically doesn't instantiate the plugin manually but normally specifies the plugin with it's string identifier. In this case I assume it's passing `adder_ripple_r25` in the hls config but I didn't check (but that's kind of the point this doesn't really give people the clues on how to use the new plugin being documented here).

mtreinish commented on releasenotes/notes/2.1/add-new-half-adder-6532be58252b2a86.yaml:
```diff
@@ -1,24 +1,27 @@
+    The :attr:`~.HalfAdderGate.definition` attribute of the :class:`.HalfAdderGate`
+    has been changed to internall use :func:`.adder_ripple_r25` to generate the definition
+    of the gate for a more efficient circuit with no ancillary qubits. If the old definition
+    is desired for some reason you can use the :func:`.adder_qft_d00` function instead which
```
```suggestion
    is desired for some reason you can directly use the :func:`.adder_qft_d00` function instead which
```

mtreinish commented on releasenotes/notes/2.1/add-new-mcx-plugins-2177db9195e2b680.yaml:
```diff
@@ -8,6 +8,6 @@ features_transpiler:
-    The :class:`.MCXSynthesisDefault` has also been updated to follow the following sequence of
-    MCX synthesizers: "mcx.2_clean_kg24", "mcx.1_clean_kg24", "mcx.n_clean_m15", "mcx.n_dirty_i15",
-    "mcx.2_dirty_kg24", "mcx.1_dirty_kg24" "mcx.1_clean_b95", "mcx.noaux_v24".
+    The :class:`.MCXSynthesisDefault` class has also been updated to follow the following sequence of
```
What does "follow the following sequence" mean here. It's not running all of them one after another, I assume it's try in this order until one succeeds, but it'd be good to be a bit more clear here.

eliarbel replied:
@alexanderivrii can you please help clarifying this?

alexanderivrii replied:
Yes, exactly, we are running the above methods one after another in the given order, until one succeeds, which in practice means there are enough clean or dirty additional auxiliary qubits available in the circuit that meet the method's requirements. Additionally, the above order in which the methods are applied is such that the better-quality methods (in terms of the number of 2-qubit gates) are before the lower-quality methods.

ShellyGarion replied:
I think it means that if you have 2 clean ancillas, then use mcx.2_clean_kg24, if you only have one clean ancilla, then use mcx.1_clean_kg24, if you only have dirty ancillas, then use mcx.n_dirty_i15 if you have n of them,  mcx.2_dirty_kg24 if you have 2 of them and mcx.1_dirty_kg24 if you only have one. If there are no ancillas then use mcx.noaux_v24.

eliarbel replied:
I've updated the explanation about the methods order. Let me know if you think it's still not clear enough.

ShellyGarion replied:
looks good!

mtreinish commented on releasenotes/notes/2.1/c-api-circuit-3ecb43d396425bed.yaml:
```diff
@@ -1,5 +1,48 @@
-    The C API for Qiskit has been extended with support for building and
-    interacting with quantum circuits.
+    The :ref:`C_API` now supports building and interacting with Quantum
+    Circuits. The circuits C API centers around the :cpp:class:`QkCircuit`
```
I checked these links I added don't resolve correctly although the formatting looks fine. We might need to update the breathe configuration to fix this, but we can look at that in a followup it's not critical for the release.

jakelishman replied:
`QkCircuit` certainly shouldn't be a `:cpp:class:` - it's not defined as a class, nor in C++.  It's probably a `` :c:struct:`QkCircuit` ``.

eliarbel replied:
It's probably because in the rst file the symbol is included using `.. doxygengroup::` and not, say, `.. doxygenstruct::`. I used `:cpp:struct:` for `QkComplex64` and it worked. I would also leave this as is, even if the cross reference does not work. 
I agree, we should fix this and possibly add custom domain roles like `:c:struct:` to Breathe.

mtreinish commented on releasenotes/notes/2.1/c-api-circuit-3ecb43d396425bed.yaml:
```diff
@@ -1,5 +1,48 @@
+    The :ref:`C_API` now supports building and interacting with Quantum
+    Circuits. The circuits C API centers around the :cpp:class:`QkCircuit`
+    opaque type that represents the circuit. The :cpp:class:`QkCircuit` type
+    enables buildign a circuit with any circuit element defined natively
```
```suggestion
    enables building a circuit with any circuit element defined natively
```

mtreinish commented on releasenotes/notes/2.1/clifford-t-optimization-cf60aad9b106820e.yaml:
```diff
@@ -2,10 +2,10 @@
+    consecutive Tdg-gates into Sdg-gates. This optimization is particularly
     effective for reducing T-count following Solovay-Kitaev decomposition, which
-    produces multiple consecutive ``T`` or ``Tdg`` gates. For example::
+    produces multiple consecutive T or Tdg gates. For example::
```
I can't comment on the note below but the same comment about `print()` in code examples.

eliarbel replied:
Added the outputs as comments to the code example.

mtreinish commented on releasenotes/notes/2.1/dag-stretch-counters-c26f39e3ea9b85a5.yaml:
```diff
@@ -1,7 +1,7 @@
 features_transpiler:
   - |
-    Adds :class:`~.DAGCircuit` attributes
+    Added :class:`~.DAGCircuit` attributes
```
A bit more detail here on what these attribute are or are used for would be good.

eliarbel replied:
Done.

mtreinish commented on releasenotes/notes/2.1/fix-assign_parameters-string-input-with-strict-false-5597010d6d0c4a5f.yaml:
```diff
@@ -0,0 +1,7 @@
+  - |
+    Fixed a bug in the :meth:`.QuantumCircuit.assign_parameters` method where parameters
+    that were not used in the circuit and were passed as strings were not ignored when
+    the argument `strict=False` was set.
```
```suggestion
    the argument ``strict=False`` was set.
```

mtreinish commented on releasenotes/notes/2.1/prepare-2.1.0-409d24ecbe277062.yaml:
```diff
@@ -0,0 +1,27 @@
+      instructions and unitary gates. Also added in this release is a set of C functions to build and manipulate a
+      :class:`.Target` object, as preparation for supporting a complete C API-based transpilation workflow in the next
+      release. In addition, ``QkComplex64`` is now defined as a struct, serving as a compiler-agnostic representation
+      for complex numbers when working with Qiskit's C API. All of the C API functions can now be used natively from C++.
```
I do kind of worry about the C++ comment here and in the other release note because it's not like we test that everything works from C++, All we did is add an `extern C` behind a c++ `ifdef` in the headers. It's fine to leave it in because if there is an issue it should be easy to fix, but I am just wondering if we want to weaken the statements here a bit.

eliarbel replied:
I relaxed the wording a little bit. Is it better now?

mtreinish replied:
I think it's better, I might tweak the wording a bit more to make it sound more like a change rather than just a note. I'll leave an inline suggestion in my next review.

mtreinish reviewed (commented):
Overall this is looking good. I had a few small inline comments nothing major.

The other thing that's missing is there are a few duplicated release notes for bugfixes that we've already release in 2.0.x release. We typically delete those duplicated release notes because the 2.0.x series goes right into 2.1.0 as the direct upgrade path. So the release notes for the 2.0.x fixes are already released and documented (this is different from 1.4.x which has a divergent history and needs to document the fixes, even if they're already released in 2.x).

There is also one release note I flagged in my local editting yesterday afternoon that I flagged as potentially needing improvement for the `PauliLindbladMap` but I didn't know what we should write there. Maybe @jakelishman or @DanPuzzuoli can help expand the content there, because the release note just lists the new classes and basically the summary line of which each class is. It doesn't really tell the user story on what the new capabilities are or what it enables.

eliarbel committed:
Address review comments

mtreinish commented on qiskit/circuit/library/arithmetic/piecewise_chebyshev.py:
```diff
@@ -377,7 +377,7 @@ class PiecewiseChebyshevGate(Gate):
             from qiskit.circuit.library.arithmetic import PiecewiseChebyshevGate
 
-            f_x, num_state_qubits, degree, breakpoints = lambda x: np.arcsin(1 / x), 2, 2, [2, 4]
+            f_x, num_state_qubits, degree, breakpoints = lambda x: np.sin(1 / x), 2, 2, [2, 4]
```
What prompted this change?

eliarbel replied:
With these parameters `PiecewiseChebyshevGate` called the lambda function with values outside the valid domain of `arcsin`, which resulted with a Sphnix warning. I thought it would just be easier to change the function to clean up the warning.

eliarbel committed:
Remove duplicate sphinx directives

mtreinish commented on releasenotes/notes/2.1/prepare-2.1.0-409d24ecbe277062.yaml:
```diff
@@ -0,0 +1,28 @@
+      :class:`.Target` object, as preparation for supporting a complete C API-based transpilation workflow in the next
+      release. In addition, :cpp:class:`QkComplex64` is now defined as a struct, serving as a compiler-agnostic representation
+      for complex numbers when working with Qiskit's C API. Note that the C API functions are generally compatible to
+      be used natively in C++.
```
```suggestion
      for complex numbers when working with Qiskit's C API. The C API's header file has also been updated so that
      it is generally compatible to be used natively in C++.
```

mtreinish reviewed (commented):
I think we're pretty close. Besides the two top level comments in my previous review about the duplicated release notes and `PauliLinbladMap` needing update I only have one small suggestion and one question inline.

eliarbel committed:
Merge remote-tracking branch 'upstream/stable/2.1' into prepare-2.1.0

jakelishman committed:
Reword PauliLindbladMap note

jakelishman commented:
I've attempted to update the `PauliLindbladMap` release note, but I'm not convinced I fully got my mathematics or the classes' intentions spot-on, so I asked Dan to have a look too.

mtreinish commented on releasenotes/notes/2.1/pauli-lindblad-map-a51b2b577c24aa95.yaml:
```diff
@@ -1,7 +1,12 @@
+    Introduced the :class:`.QubitSparsePauli` and :class:`.QubitSparsePauliList` classes, which
+    represent the same concepts as :class:`.Pauli` and :class:`.PauliList` respectively, but only
+    store non-identity terms, in a manner analogous to :class:`.SparseObservable`.  These classes
+    are primarily intended to be used with teh new :class:`.PauliLindbladMap`.
```
```suggestion
    are primarily intended to be used with the new :class:`.PauliLindbladMap`.
```

jakelishman replied:
Done in 886d48f57b.

mtreinish commented on releasenotes/notes/2.1/pauli-lindblad-map-a51b2b577c24aa95.yaml:
```diff
@@ -1,7 +1,12 @@
\ No newline at end of file
+    A new class, :class:`.PauliLindbladMap`, is added, which is a Pauli-based parametrization of
+    a subset of quantum channels, used in noise-learning applications.  This class is expected to
+    form the backbone of enhanced noise-learning algorithms, and provide better and more effiicent
```
```suggestion
    form the backbone of enhanced noise-learning algorithms, and provide better and more efficient
```

jakelishman replied:
Doenin 886d48f57b.

jakelishman committed:
Fixup PauliLindbladMap wording

jakelishman commented:
Dan looked at the release note, and we reworded it to what you see in 886d48f57b.

eliarbel committed:
Update releasenotes/notes/2.1/prepare-2.1.0-409d24ecbe277062.yaml

Co-authored-by: Matthew Treinish <mtreinish@kortar.org>

eliarbel committed:
Remove duplicate release notes from 2.0.x

eliarbel committed:
Merge branch 'prepare-2.1.0' of github.com:eliarbel/qiskit into prepare-2.1.0

eliarbel committed:
Resolve Pauli cross-reference in a release note

mtreinish committed:
Update fix notes

mtreinish committed:
Merge branch 'stable/2.1' into prepare-2.1.0

mtreinish reviewed (approved):
I think this LGTM now. Thanks for all the work on this.

I just pushed up some small final edits and removals of published release notes. reading through the rendered notes I don't see anything else that looks out of place to me (that being said given the volume of text I'm probably missing something but we can always fix those post release as part of a 2.1.1)
